src/compiler.ts

```diff
@@ -6512,14 +6598,64 @@ export class Compiler extends DiagnosticEmitter {
         return module.unreachable();
       }
     }
-    return this.compileCallIndirect(
-      assert(signature), // FIXME: asc can't see this yet
-      indexArg,
-      expression.arguments,
-      expression,
-      0,
-      contextualType == Type.void
-    );
+
+    var returnType = signature.returnType;
+    var tempResultLocal = this.currentFlow.getTempLocal(returnType);
+    var tempFunctionReferenceLocal = this.currentFlow.getTempLocal(returnType);
+
+    var callExpr = module.block(null, [
+      module.local_set(
+        tempFunctionReferenceLocal.index,
+        indexArg
+      ),
+      module.push(
+        module.global_get(BuiltinNames.global_closure, NativeType.I32)
+      ),
+      this.compileCallIndirect(
+        assert(signature), // FIXME: asc can't see this yet
+        module.if(
+          module.binary(
+            BinaryOp.EqI32,
+            module.binary(
+              BinaryOp.AndI32,
+              module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+              module.i32(1 << 31)
+            ),
+            module.i32(1 << 31)
+          ),
+          module.block(null, [
+            module.global_set(
+              BuiltinNames.global_closure,
+              module.binary(
+                BinaryOp.ShlI32,
+                module.local_get(tempFunctionReferenceLocal.index, NativeType.I32),
+                module.i32(4)
+              )
+            ),
+            module.load( //TODO: support 8 byte addrs
+              4,
+              true,
+              module.global_get(BuiltinNames.global_closure, NativeType.I32),
+              NativeType.I32,
+              0
+            )
+          ], this.options.nativeSizeType),
```

Comment:
If it's a closure, move the closure context into the global closure context, then access the first field, which is always the function index. Otherwise we just put the index argument as-is onto the stack